vuvuzela_test.go

```diff
@@ -0,0 +1,304 @@
+		Handler:            u.Handler,
+		CoordinatorLatency: 150 * time.Millisecond,
+	}
+	err := client.Bootstrap(u.ConvoConfig)
+	if err != nil {
+		log.Panicf("bootstrapping vuvuzela client: %s", err)
+	}
+	err = client.Persist()
+	if err != nil {
+		log.Panicf("persisting vuvuzela client: %s", err)
+	}
+	return client
+}
+
+type mockHandler struct {
+	errPrefix      string
+	sentReplies    chan roundAndReplies
+	newConfig      chan []*config.SignedConfig
+	sentOutgoing   chan roundAndDeadDropMsgs
+	agreedDeadDrop *convo.DeadDrop
+}
+
+type roundAndReplies struct {
+	round   uint32
+	replies [][]byte
+}
+
+type roundAndDeadDropMsgs struct {
+	round    uint32
+	messages []*convo.DeadDropMessage
+}
+
+// Outgoing sends a random message to the agreed dead drop.
+func (h *mockHandler) Outgoing(round uint32) []*convo.DeadDropMessage {
+	out := make([]*convo.DeadDropMessage, 0, 1)
+	msg := new(convo.DeadDropMessage)
+	msg.DeadDrop = *h.agreedDeadDrop
+	rand.Read(msg.EncryptedMessage[:])
+	out = append(out, msg)
+	h.sentOutgoing <- roundAndDeadDropMsgs{round: round, messages: out}
+	return out
+}
+
+func (h *mockHandler) Replies(round uint32, replies [][]byte) {
+	h.sentReplies <- roundAndReplies{round: round, replies: replies}
+}
+func (h *mockHandler) NewConfig(chain []*config.SignedConfig) {
+	h.newConfig <- chain
+}
+func (h *mockHandler) Error(err error) {
+	log.Errorf(h.errPrefix+": client error: %s", err)
+}
+func (h *mockHandler) DebugError(err error) {
+	log.Errorf(h.errPrefix+": debug client error: %s", err)
+}
+func (h *mockHandler) GlobalAnnouncement(message string) {
+	log.Debugf("GLOBAL ANNOUNCEMENT: %s", message)
+}
+
+func TestVuvuzela(t *testing.T) {
+	u := createVuvuzelaUniverse()
+	defer func() {
+		// Give time for goroutines to finish before pulling the rug out from under them
+		time.Sleep(1 * time.Second)
+		u.Destroy()
+	}()
+
+	aliceClient := u.createUser("alice")
+	_, err := aliceClient.ConnectConvo()
+	if err != nil {
+		t.Fatalf("Failed to connect Alice's conversation: %s", err)
+	}
+	defer aliceClient.CloseConvo()
+
+	bobClient := u.createUser("bob")
+	_, err = bobClient.ConnectConvo()
+	if err != nil {
+		t.Fatalf("Failed to connect Bob's conversation: %s", err)
+	}
+	defer bobClient.CloseConvo()
+
+	aliceRoundAndDeadDropMsgs := <-aliceClient.Handler.(*mockHandler).sentOutgoing
+	aliceRound := aliceRoundAndDeadDropMsgs.round
+	aliceMsgs := aliceRoundAndDeadDropMsgs.messages
+
+	bobRoundAndDeadDropMsgs := <-bobClient.Handler.(*mockHandler).sentOutgoing
+	bobRound := bobRoundAndDeadDropMsgs.round
+	bobMsgs := bobRoundAndDeadDropMsgs.messages
+
+	aliceRoundAndReplies := <-aliceClient.Handler.(*mockHandler).sentReplies
+	aliceReplyRound := aliceRoundAndReplies.round
+	aliceReplies := aliceRoundAndReplies.replies
+
+	bobRoundAndReplies := <-bobClient.Handler.(*mockHandler).sentReplies
+	bobReplyRound := bobRoundAndReplies.round
+	bobReplies := bobRoundAndReplies.replies
+
+	if len(bobMsgs) != 1 {
```

Comment:
Replace these lines with checks that Alice got Bob's message and vice-versa.

Reply:
Done